Source/WebCore/workers/service/ServiceWorkerRoute.cpp

```diff
@@ -0,0 +1,261 @@
+
+    if (!matchURLPatternComponent(urlPattern.username, url.encodedUser()))
+        return false;
+
+    if (!matchURLPatternComponent(urlPattern.password, url.encodedPassword()))
+        return false;
+
+    if (!matchURLPatternComponent(urlPattern.hostname, url.host()))
+        return false;
+
+    String port;
+    if (auto portNumber = url.port())
+        port = String::number(*portNumber);
+    if (!matchURLPatternComponent(urlPattern.port, port))
+        return false;
+
+    if (!matchURLPatternComponent(urlPattern.pathname, url.path()))
+        return false;
+
+    if (!matchURLPatternComponent(urlPattern.search, url.query()))
+        return false;
+
+    return matchURLPatternComponent(urlPattern.hash, url.fragmentIdentifier());
+}
+
+// https://w3c.github.io/ServiceWorker/#match-router-condition
+bool matchRouterCondition(const ServiceWorkerRouteCondition& condition, const FetchOptions& options, const ResourceRequest& request, bool isServiceWorkerRunning)
+{
+    if (!condition.orConditions.isEmpty()) {
+        for (auto& condition : condition.orConditions) {
+            if (matchRouterCondition(condition, options, request, isServiceWorkerRunning))
+                return true;
+        }
+        return false;
+    }
+
+    if (condition.notCondition)
+        return !matchRouterCondition(*condition.notCondition, options, request, isServiceWorkerRunning);
+
+    if (condition.urlPattern) {
+        if (!matchURLPattern(*condition.urlPattern, request.url()))
+            return false;
+    }
+
+    if (!condition.requestMethod.isNull()) {
+        if (condition.requestMethod != request.httpMethod())
+            return false;
+    }
+
+    if (condition.requestMode) {
+        if (*condition.requestMode != options.mode)
+            return false;
+    }
+
+    if (condition.requestDestination) {
+        if (*condition.requestDestination != options.destination)
+            return false;
+    }
+
+    if (condition.runningStatus) {
+        bool isRunningStatus = *condition.runningStatus == RunningStatus::Running;
+        if (isRunningStatus != isServiceWorkerRunning)
+            return false;
+    }
+
+    return true;
+}
+
+ServiceWorkerRouteCondition ServiceWorkerRouteCondition::isolatedCopy() &&
+{
+    std::unique_ptr<ServiceWorkerRouteCondition> notConditionCopy;
+    if (notCondition)
+        notConditionCopy = makeUnique<ServiceWorkerRouteCondition>(WTFMove(*notCondition));
+    return {
+        crossThreadCopy(WTFMove(urlPattern)),
+        crossThreadCopy(WTFMove(requestMethod)),
+        requestMode,
+        requestDestination,
+        runningStatus,
+        crossThreadCopy(WTFMove(orConditions)),
+        WTFMove(notConditionCopy)
+    };
+}
+
+ServiceWorkerRouteCondition ServiceWorkerRouteCondition::copy() const
+{
+    std::unique_ptr<ServiceWorkerRouteCondition> notConditionCopy;
+    if (notCondition)
+        notConditionCopy = makeUnique<ServiceWorkerRouteCondition>(WTFMove(*notCondition));
```

Comment:
How does moving `*notCondition` even work in this const member function? Maybe I am misreading it.

Reply:
WTFMove is indeed wrong, I'll remove it.